runtime/lua/vim/diagnostic.lua

```diff
@@ -1857,16 +1857,37 @@ function M.open_float(opts, ...)
   if scope == 'line' then
     --- @param d vim.Diagnostic
     diagnostics = vim.tbl_filter(function(d)
-      return lnum >= d.lnum and lnum <= d.end_lnum
+      if lnum < d.lnum or lnum > d.end_lnum then
+        return false
+      end
+      if d.lnum == d.end_lnum then
+        -- if diagnostic only spans one line, always show
+        -- allows showing diagnostics with zero width (where lnum == end_lnum and col == end_col)
+        return true
+      else
+        -- end-exclusive range, don't show if on end_lnum and end_col == 0
+        return lnum ~= d.end_lnum or d.end_col ~= 0
+      end
```

Comment:
It looks like the logic here is the same as before, except for the addition of the end-exclusive range. I find the refactored logic here hard to follow (multiple returns and if statements), we can keep the original structure and just add the new condition:

```suggestion
      return (lnum >= d.lnum and lnum <= d.end_lnum) and (lnum < d.end_lnum or d.end_col ~= 0)

```

Reply:
That condition isn't the same - if the diagnostic is `{ lnum = 0, end_lnum = 0, col = 0, end_col = 0 }` (zero width) then that will return `false` when the cursor is on line 0, meaning the diagnostic will never be "hoverable".

Reply:
I did make an edit, so it should be the same now. I ran it against your test cases and they all pass.

Reply:
Originally I tried to keep both branches as one big conditional statement, but the `scope = 'cursor'` case was too complex for that, so I decided to change both to use if-then-returns for consistency. But it's certainly not a big deal - I don't mind changing it.